File: Inventarios.Api.Nest/productos.api/src/domain/dtos/responses/api.response.ts
export class ApiResponse<T> {
  success: boolean;
  message: string;
  data?: T;
  errors?: string[];

  private constructor(success: boolean, message: string, data?: T, errors?: string[]) {
    this.success = success;
    this.message = message;
    this.data = data;
    this.errors = errors;
  }

  static ok<T>(data: T, message = 'OK'): ApiResponse<T> {
    return new ApiResponse(true, message, data);
  }

  static fail<T>(message: string, errors?: string[]): ApiResponse<T> {
    return new ApiResponse<T>(false, message, undefined, errors);
  }
}

File: Inventarios.Api/Inventarios.api.Domain/dtos/responses/ApiResponse.cs
namespace Inventarios.api.Domain.Dtos.Responses;

/// <summary>
/// Envelope estándar para todas las respuestas de la API.
/// Garantiza consistencia en el contrato HTTP.
/// </summary>
public sealed record ApiResponse<T>(
    bool Success,
    string Message,
    T? Data,
    IEnumerable<string>? Errors = null
)
{
    public static ApiResponse<T> Ok(T data, string message = "OK") =>
        new(true, message, data);

    public static ApiResponse<T> Fail(string message, IEnumerable<string>? errors = null) =>
        new(false, message, default, errors);
}